tools/clang/lib/CodeGen/CGClass.cpp

```diff
@@ -1768,16 +1780,31 @@ void CodeGenFunction::EmitCXXAggrConstructorCall(const CXXConstructorDecl *ctor,
                            /*Delegating=*/false, cur, E);
   }
 
-  // Go to the next element.
-  llvm::Value *next =
-    Builder.CreateInBoundsGEP(cur, llvm::ConstantInt::get(SizeTy, 1),
-                              "arrayctor.next");
-  cur->addIncoming(next, Builder.GetInsertBlock());
+  // HLSL Change Begin: Loop on index instead of ptr
+  //// Go to the next element.
+  //llvm::Value *next =
+  //  Builder.CreateInBoundsGEP(cur, llvm::ConstantInt::get(SizeTy, 1),
+  //                            "arrayctor.next");
+  //cur->addIncoming(next, Builder.GetInsertBlock());
+  idx->addIncoming(next, Builder.GetInsertBlock());
+  // HLSL Change End
 
   // Check whether that's the end of the loop.
-  llvm::Value *done = Builder.CreateICmpEQ(next, arrayEnd, "arrayctor.done");
+  // HLSL Change Begin: Loop on index instead of ptr
+  //llvm::Value *done = Builder.CreateICmpEQ(next, arrayEnd, "arrayctor.done");
+  llvm::Value *done = Builder.CreateICmpEQ(next, numElements, "arrayctor.done");
+  // HLSL Change End
+
   llvm::BasicBlock *contBB = createBasicBlock("arrayctor.cont");
-  Builder.CreateCondBr(done, contBB, loopBB);
+  llvm::TerminatorInst *TI = cast<llvm::TerminatorInst>(  // HLSL Change, capture terminator
+    Builder.CreateCondBr(done, contBB, loopBB));
+
+  // HLSL Change Begin: force unroll
```

Comment:
Without the attribute, unroll is not attempted with -Od, or is otherwise dependent on default heuristics.  Without unroll, SROA would fail, before the indexing change.  With the indexing change, this should technically work without unrolling, however, unrolling is required to potentially eliminate unused expensive RayQuery allocations.  Since the loop bounds should be constant (for us), and unrolling/DCE should not hurt debugging here, there's no reason we shouldn't unroll initialization loops.